File: services/savings-service/utils/coa_client.py
"""
Chart of Accounts Client for Savings Service

Add this to utils/coa_client.py or import from payment-processing-service
"""

import os
import logging
from typing import Dict, Any, Optional
from datetime import datetime
import httpx

logger = logging.getLogger(__name__)


class CoAClient:
    """Client for Chart of Accounts service"""
    
    def __init__(self, base_url: Optional[str] = None):
        self.base_url = base_url or os.getenv("COA_SERVICE_URL", "http://chart-of-accounts-service:8080")
        self.timeout = 30.0
    
    def _get_headers(self, tenant_id: str, user_id: str, user_role: str) -> Dict[str, str]:
        return {
            "Content-Type": "application/json",
            "X-Tenant-ID": tenant_id,
            "x-keycloak-id": user_id,
            "X-User-Role": user_role,
        }
    
    async def record_savings_deposit(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        account_id: str,
        amount: int,  # Amount in kobo
        customer_account: str,
    ) -> Dict[str, Any]:
        """
        Record savings deposit
        
        Debit: Savings Deposits (2200)
        Credit: Customer source account
        """
        entry_data = {
            "date": datetime.utcnow().isoformat(),
            "description": f"Savings deposit to account {account_id}",
            "posted_by": user_id,
            "reference": account_id,
            "lines": [
                {
                    "account_id": "2200",  # Savings Deposits liability
                    "description": f"Deposit to savings account {account_id}",
                    "debit_amount": amount,
                    "credit_amount": 0,
                },
                {
                    "account_id": customer_account,
                    "description": "Transfer from customer account",
                    "debit_amount": 0,
                    "credit_amount": amount,
                },
            ],
            "metadata": {
                "account_id": account_id,
                "source": "savings-service",
                "event_type": "deposit",
            },
        }
        
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            try:
                response = await client.post(
                    f"{self.base_url}/api/v1/journal-entries",
                    json=entry_data,
                    headers=self._get_headers(tenant_id, user_id, user_role),
                )
                response.raise_for_status()
                return response.json()
            except httpx.HTTPError as e:
                logger.error(f"Failed to record savings deposit: {e}")
                raise
    
    async def record_savings_withdrawal(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        account_id: str,
        amount: int,
        customer_account: str,
    ) -> Dict[str, Any]:
        """
        Record savings withdrawal
        
        Debit: Customer destination account
        Credit: Savings Deposits (2200)
        """
        entry_data = {
            "date": datetime.utcnow().isoformat(),
            "description": f"Savings withdrawal from account {account_id}",
            "posted_by": user_id,
            "reference": account_id,
            "lines": [
                {
                    "account_id": customer_account,
                    "description": "Withdrawal to customer account",
                    "debit_amount": amount,
                    "credit_amount": 0,
                },
                {
                    "account_id": "2200",  # Savings Deposits
                    "description": f"Withdrawal from savings account {account_id}",
                    "debit_amount": 0,
                    "credit_amount": amount,
                },
            ],
            "metadata": {
                "account_id": account_id,
                "source": "savings-service",
                "event_type": "withdrawal",
            },
        }
        
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            try:
                response = await client.post(
                    f"{self.base_url}/api/v1/journal-entries",
                    json=entry_data,
                    headers=self._get_headers(tenant_id, user_id, user_role),
                )
                response.raise_for_status()
                return response.json()
            except httpx.HTTPError as e:
                logger.error(f"Failed to record savings withdrawal: {e}")
                raise
    
    async def record_interest_accrual(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        account_id: str,
        interest_amount: int,
    ) -> Dict[str, Any]:
        """
        Record interest accrual on savings
        
        Debit: Interest Expense (5200)
        Credit: Savings Deposits (2200)
        """
        entry_data = {
            "date": datetime.utcnow().isoformat(),
            "description": f"Interest accrual for savings account {account_id}",
            "posted_by": user_id,
            "reference": account_id,
            "lines": [
                {
                    "account_id": "5200",  # Interest Expense
                    "description": "Interest expense on savings",
                    "debit_amount": interest_amount,
                    "credit_amount": 0,
                },
                {
                    "account_id": "2200",  # Savings Deposits
                    "description": f"Interest credited to account {account_id}",
                    "debit_amount": 0,
                    "credit_amount": interest_amount,
                },
            ],
            "metadata": {
                "account_id": account_id,
                "source": "savings-service",
                "event_type": "interest_accrual",
            },
        }
        
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            try:
                response = await client.post(
                    f"{self.base_url}/api/v1/journal-entries",
                    json=entry_data,
                    headers=self._get_headers(tenant_id, user_id, user_role),
                )
                response.raise_for_status()
                return response.json()
            except httpx.HTTPError as e:
                logger.error(f"Failed to record interest accrual: {e}")
                raise


# Integration example in savings service
"""
from utils.coa_client import CoAClient

class SavingsService:
    def __init__(self):
        self.coa_client = CoAClient()
    
    async def process_deposit(self, account_id, amount, customer_account, context):
        # Record in Chart of Accounts (fail-fast)
        try:
            await self.coa_client.record_savings_deposit(
                tenant_id=context.tenant_id,
                user_id=context.user_id,
                user_role="bank_admin",
                account_id=account_id,
                amount=amount,  # in kobo
                customer_account=customer_account,
            )
        except Exception as e:
            logger.error(f"Failed to record journal entry for deposit: {e}")
            raise Exception("Failed to record accounting entry. Deposit not processed.")
        # ... process the deposit in your system ...
    
    async def process_withdrawal(self, account_id, amount, customer_account, context):
        # Record in Chart of Accounts (fail-fast)
        try:
            await self.coa_client.record_savings_withdrawal(
                tenant_id=context.tenant_id,
                user_id=context.user_id,
                user_role="bank_admin",
                account_id=account_id,
                amount=amount,
                customer_account=customer_account,
            )
        except Exception as e:
            logger.error(f"Failed to record journal entry for withdrawal: {e}")
            raise Exception("Failed to record accounting entry. Withdrawal not processed.")
        # ... process the withdrawal in your system ...
"""

File: services/stk-service/coa_client.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

// CoAClient is a client for interacting with the Chart of Accounts service
type CoAClient struct {
	baseURL    string
	httpClient *http.Client
}

// NewCoAClient creates a new Chart of Accounts client
func NewCoAClient() *CoAClient {
	baseURL := os.Getenv("COA_SERVICE_URL")
	if baseURL == "" {
		baseURL = "http://chart-of-accounts-service:8080"
	}

	return &CoAClient{
		baseURL: baseURL,
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// CoAAccount represents an account in the Chart of Accounts
type CoAAccount struct {
	ID              string    `json:"id,omitempty"`
	TenantID        string    `json:"tenant_id"`
	Code            string    `json:"code"`
	Name            string    `json:"name"`
	Type            string    `json:"type"` // asset, liability, equity, revenue, expense
	ParentID        string    `json:"parent_id,omitempty"`
	NormalBalance   string    `json:"normal_balance,omitempty"`
	Level           int       `json:"level,omitempty"`
	IsActive        bool      `json:"is_active"`
	IsSystemAccount bool      `json:"is_system_account"`
	Currency        string    `json:"currency"`
	Description     string    `json:"description,omitempty"`
	CreatedAt       time.Time `json:"created_at,omitempty"`
	UpdatedAt       time.Time `json:"updated_at,omitempty"`
}

// JournalLineRequest represents a single line in a journal entry
type JournalLineRequest struct {
	AccountID    string `json:"account_id"` // Can be account code or UUID
	Description  string `json:"description,omitempty"`
	DebitAmount  int64  `json:"debit_amount"`  // Amount in smallest currency unit (e.g., kobo for NGN)
	CreditAmount int64  `json:"credit_amount"` // Amount in smallest currency unit
}

// CreateJournalEntryRequest represents a request to create a journal entry
type CreateJournalEntryRequest struct {
	Date        time.Time              `json:"date"`
	Description string                 `json:"description"`
	Reference   string                 `json:"reference,omitempty"`
	Lines       []JournalLineRequest   `json:"lines"`
	PostedBy    string                 `json:"posted_by"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
}

// JournalEntryResponse represents the response from creating a journal entry
type JournalEntryResponse struct {
	ID          string                 `json:"id"`
	TenantID    string                 `json:"tenant_id"`
	EntryNumber string                 `json:"entry_number"`
	Date        time.Time              `json:"date"`
	Description string                 `json:"description"`
	Status      string                 `json:"status"`
	Lines       []JournalLine          `json:"lines"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
	CreatedAt   time.Time              `json:"created_at"`
	UpdatedAt   time.Time              `json:"updated_at"`
}

// JournalLine represents a line in a journal entry
type JournalLine struct {
	ID           string `json:"id"`
	AccountID    string `json:"account_id"`
	AccountCode  string `json:"account_code"`
	AccountName  string `json:"account_name"`
	Description  string `json:"description"`
	DebitAmount  int64  `json:"debit_amount"`
	CreditAmount int64  `json:"credit_amount"`
}

// CreateAccount creates a new account in the Chart of Accounts
func (c *CoAClient) CreateAccount(tenantID, userID, userRole string, account CoAAccount) (*CoAAccount, error) {
	account.TenantID = tenantID

	body, err := json.Marshal(account)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal account: %w", err)
	}

	req, err := http.NewRequest("POST", c.baseURL+"/api/v1/accounts", bytes.NewBuffer(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Tenant-ID", tenantID)
	req.Header.Set("x-keycloak-id", userID)
	req.Header.Set("X-User-Role", userRole)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("failed to create account: %s (status: %d)", string(bodyBytes), resp.StatusCode)
	}

	var result CoAAccount
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result, nil
}

// CreateJournalEntry creates a new journal entry
func (c *CoAClient) CreateJournalEntry(tenantID, userID, userRole string, entry CreateJournalEntryRequest) (*JournalEntryResponse, error) {
	body, err := json.Marshal(entry)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal journal entry: %w", err)
	}

	req, err := http.NewRequest("POST", c.baseURL+"/api/v1/journal-entries", bytes.NewBuffer(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Tenant-ID", tenantID)
	req.Header.Set("x-keycloak-id", userID)
	req.Header.Set("X-User-Role", userRole)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("failed to create journal entry: %s (status: %d)", string(bodyBytes), resp.StatusCode)
	}

	var result JournalEntryResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result, nil
}

// GetAccounts retrieves all accounts for a tenant
func (c *CoAClient) GetAccounts(tenantID, userID, userRole string) ([]CoAAccount, error) {
	req, err := http.NewRequest("GET", c.baseURL+"/api/v1/accounts", nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("X-Tenant-ID", tenantID)
	req.Header.Set("x-keycloak-id", userID)
	req.Header.Set("X-User-Role", userRole)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("failed to get accounts: %s (status: %d)", string(bodyBytes), resp.StatusCode)
	}

	var accounts []CoAAccount
	if err := json.NewDecoder(resp.Body).Decode(&accounts); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return accounts, nil
}

// RecordLoanDisbursement creates a journal entry for loan disbursement
// Debit: Loans Receivable (1400)
// Credit: Customer Account (liability account)
func (c *CoAClient) RecordLoanDisbursement(tenantID, userID, userRole, loanID string, amount int64, customerAccountCode string) (*JournalEntryResponse, error) {
	entry := CreateJournalEntryRequest{
		Date:        time.Now(),
		Description: fmt.Sprintf("Loan disbursement for loan %s", loanID),
		Reference:   loanID,
		PostedBy:    userID,
		Lines: []JournalLineRequest{
			{
				AccountID:    "1400", // Loans Receivable
				Description:  "Loan principal disbursed",
				DebitAmount:  amount,
				CreditAmount: 0,
			},
			{
				AccountID:    customerAccountCode,
				Description:  "Credit to customer account",
				DebitAmount:  0,
				CreditAmount: amount,
			},
		},
		Metadata: map[string]interface{}{
			"loan_id":    loanID,
			"source":     "loan-service",
			"event_type": "disbursement",
		},
	}

	return c.CreateJournalEntry(tenantID, userID, userRole, entry)
}

// RecordLoanRepayment creates a journal entry for loan repayment
// Debit: Customer Account (or Cash)
// Credit: Loans Receivable (1400)
func (c *CoAClient) RecordLoanRepayment(tenantID, userID, userRole, loanID string, principalAmount, interestAmount int64, sourceAccountCode string) (*JournalEntryResponse, error) {
	totalAmount := principalAmount + interestAmount

	lines := []JournalLineRequest{
		{
			AccountID:    sourceAccountCode,
			Description:  "Loan repayment received",
			DebitAmount:  totalAmount,
			CreditAmount: 0,
		},
		{
			AccountID:    "1400", // Loans Receivable
			Description:  "Principal repayment",
			DebitAmount:  0,
			CreditAmount: principalAmount,
		},
	}

	// If there's interest, add a credit to interest revenue account
	if interestAmount > 0 {
		lines = append(lines, JournalLineRequest{
			AccountID:    "4100", // Interest Revenue
			Description:  "Interest earned",
			DebitAmount:  0,
			CreditAmount: interestAmount,
		})
	}

	entry := CreateJournalEntryRequest{
		Date:        time.Now(),
		Description: fmt.Sprintf("Loan repayment for loan %s", loanID),
		Reference:   loanID,
		PostedBy:    userID,
		Lines:       lines,
		Metadata: map[string]interface{}{
			"loan_id":          loanID,
			"source":           "loan-service",
			"event_type":       "repayment",
			"principal_amount": principalAmount,
			"interest_amount":  interestAmount,
		},
	}

	return c.CreateJournalEntry(tenantID, userID, userRole, entry)
}
